---
- name: "Checking {{ course_name }} lab {{ lab_num }}"
  hosts: all
  become: true
  gather_facts: false
  ignore_errors: yes
  vars:
    failed_test_reason: []
    failed_test_counter: 0
    # test_counter: 0

  tasks:

    - name: Check mode block - don't fix student errors
      block:

        - name: Section 2.1 Install Ansible
          block:

            - name: Lab02 Section 2.1 Check Ansible installed
              yum:
                name: ansible
                state: present
              register: last_task

            - name: Log failed test
              set_fact:
                failed_test_reason: "{{ failed_test_reason }} + ['Section 2.1: Ansible package not present on bastion']"
                failed_test_counter: "{{ failed_test_counter | int + 1 }}"
              delegate_to: localhost  
              delegate_facts: True
              when: last_task.rc == 0
          when: inventory_hostname in groups['bastions']   

        - name: Debugging structure for outputting current status
          debug:
            msg: 
              - "Failed Task Reason: {{ failed_test_reason }}"
              - "Number of failed tests: {{ failed_test_counter }}"  
            verbosity: 2    
          delegate_to: localhost      
          when: inventory_hostname == "localhost"
        - name: Section 2.2.1 Create user devops
          block:

            - name: Lab02 Section 2.2.1  Does user devops exist
              user:
                name: devops
                state: present
            - name: Log failed test
              set_fact:
                failed_test_reason: "{{hostvars['localhost']['failed_test_reason']}} + ['Section 2.2.1: User devops not present on bastion']"
                failed_test_counter: "{{ hostvars['localhost']['failed_test_counter'] | int + 1 }}"
              when: last_task.rc != 1
              delegate_to: localhost  
              delegate_facts: True
          when: inventory_hostname in groups['bastions']   


        - name: Debugging structure for outputting current status
          debug:
            msg: 
              - "Failed Task Reason: {{ failed_test_reason }}"
              - "Number of failed tests: {{ failed_test_counter }}"  
            verbosity: 2    
          delegate_to: localhost      
          when: inventory_hostname == "localhost"

        - name: Section 2.2.3 Create user devops
          block:

            - name: Lab02 Section 2.2 Does ssh key exist for user devops
              stat:
                path: /home/devops/.ssh/id_rsa.pub
              register: key_stat  

            - name: Log failed test
              set_fact:
                failed_test_reason: "{{ failed_test_reason }} + ['Section 2.2.3: User devops ssh key not present on bastion']"
                failed_test_counter: "{{ failed_test_counter | int + 1 }}"   
              when: key_stat.stat.exists == False
          when: inventory_hostname in groups['bastions']   

        - name: Lab02 Section 2.2 Should fail - FAKE Does ssh key exist for user devops
          stat:
            path: /home/devops/.ssh/iiiiiid_rsa.pub
          register: key_stat  
          changed_when: key_stat.stat.exists == False

      vars:    
        ansible_check_mode: True
...
